### ISSUE TITLE
Faster FFMPEG Concept

### ISSUE THREAD

**Issue by akai-katto** [k4yt3x/video2x#91]
I have an idea for Dandere2x that would also be applicable to Video2x (I haven't programmed it yet since busy) and I was wondering if you'd be interested in looking over the proposal and are interested in potentially contributing to it. 

https://github.com/CardinalPanda/dandere2x/wiki/Proposed-Method-of-Reducing-Overall-Runtime-in-Video-Upscaling

(i'm not really sure if this is an inappropriate place to submit a suggestion like this, let me know if it is.)

**Comment by k4yt3x** [comment]
@CardinalPanda looked at the proposal. It makes sense, but the FFmpeg part usually just takes seconds to finish, it's not the bottleneck of upscaling really. I am worried that implementing this will result in more-complex and harder-to-maintain code.

I'll do some searches on this topic, but I don't think it should be prioritized.

> (i'm not really sure if this is an inappropriate place to submit a suggestion like this, let me know if it is.)

This is exactly where it should be.

**Comment by CardinalPanda** [comment]
> This is exactly where it should be.

Thanks for letting me know!

I get around ~7-14fps when encoding 4k frames into a video. This motivation came up when it took roughly half an hour to encode 20k (13 minutes @ 24fps) frames into a viewable format, even slower if I choose to compress the video using some codec (5-7 fps).

**Comment by k4yt3x** [comment]
This might depend on the CPU I guess. My Ryzen 7 1700 encodes 10 minute videos in seconds.
I'll take a look at the concatenation solution, wonder if I can do that with 264 or 265.

**Comment by Inrixia** [comment]
Sorry to bump a stale thread, but this is actually something that would really help with my upscale times. Currently when upscaling decoding a video to frames and then encoding can take as long if not longer than the video runtime on my 6700k so having a way to mitigate that would drastically speed up the process, especially when we are talking about up scaling hours of content.

Even just a way to have it pre-process frames when running in bulk mode would be good. I have found that if you set the wifu2x process as high priority while ffmpeg is running even though the cpu is maxed you get basically no performance hit on the upscale process. So if it was possible to always have frames ready to upscale and avoid having gpu's sitting idle waiting for ffmpeg that would be great.

**Comment by k4yt3x** [comment]
@Inrixia acknowledged. This though, seems like a lot of changes to the current FFmpeg code.
I'll have to try to dedicate a long section of time to finish this in one run.

**Comment by aka-katto** [comment]
Dandere2x uses progressive frame extraction (pauses / resumes ffmpeg when it needs more frames) and piping to address this same issue. 

Video2x behaves pretty differently than d2x, but if you need help with ideas / implementation, im pretty experienced / have done work for this specific problem

**Comment by k4yt3x** [comment]
> Dandere2x uses progressive frame extraction (pauses / resumes ffmpeg when it needs more frames) and piping to address this same issue.

I'm very interested in how this works. This seems like an ideal solution.

**Comment by aka-katto** [comment]
> 
> 
> > Dandere2x uses progressive frame extraction (pauses / resumes ffmpeg when it needs more frames) and piping to address this same issue.
> 
> I'm very interested in how this works. This seems like an ideal solution.

i'll write a brief white paper of the tools we used to get it to work so you can see what would need to change for video2x to work with it

**Comment by k4yt3x** [comment]
@aka-katto much appreciated

**Comment by aka-katto** [comment]
it's a bit brief, but here's some notes that I think should give you an idea how dandere2x works now. Let me know if anything is unclear 

# Abstract

This is a condensed set of notes expressing how dandere2x can keep the number of images written on a disk to a minimum. 

*Credit to Tremex for conceptualizing/developing this.*


# Methods

## Progressive Frame Extractor

Code: https://github.com/aka-katto/dandere2x/blob/master/src/wrappers/ffmpeg/progressive_frame_extractor_ffmpeg.py

This feature works primarily through creating an FFmpeg subprocess, then using the psutil library to suspend the subprocess as needed, and un-freezing the process when more frames are needed. 


## Image Piping

Code: https://github.com/aka-katto/dandere2x/blob/master/src/wrappers/ffmpeg/pipe.py

ffmpeg yaml:
```
  pipe_video:
    -y: true
    output_options:
      -loglevel: panic
      -y: -f
      image2pipe: True
      -i: '-'
      -vcodec: libx264
      -preset: medium
      -qscale: 5
      -crf: 17
      -vf:
        - ' pp=hb/vb/dr/fq|32, deband=range=22:blur=false'
```

Relevant Resource: http://zulko.github.io/blog/2013/09/27/read-and-write-video-frames-in-python-using-ffmpeg/

Images are directly piped into ffmpeg, where it begins to get processed immediately (i assume), rather than after the upscale process is completed. I'm not sure exactly where the images are saved to when buffered in, but it seems to eliminate the need for having all N files saved to disk to create the video since once an image is piped, it can be deleted.

# Current Issues

- If dandere2x finishes before progressive_frame_extractor is called, a suspending ffmpeg.exe will run in the background. This can easily be fixed by manually killing the subprocess on exit.

- Piping uses a lot of ram, tremex noted that vp9 uses significantly less than h264, but I've left 264 to be the default due to uniformity.

**Comment by k4yt3x** [comment]
Implemented in 5.0.0
